neo/Network/P2P/ProtocolHandler.cs

```diff
@@ -253,7 +253,7 @@ private void OnInventoryReceived(IInventory inventory)
 
         private void OnInvMessageReceived(InvPayload payload)
         {
-            UInt256[] hashes = payload.Hashes.Where(p => knownHashes.Add(p)).ToArray();
+            UInt256[] hashes = payload.Hashes.Where(p => knownHashes.Add(p) && !sentHashes.Contains(p)).ToArray();
```

Comment:
This will naturally add to knownHashes, as Add will run before sentHashes is verified... interesting.

Reply:
The intend is to verify this:
```cs
            bool sentByMe = payload.Hashes.Any(p => sentHashes.Contains(p));
            if (hashes.Length == 0 || sentByMe) return;
```

Check if there is Any of this hashes in the SentHashes.

Is this the same as it is? Shargon suggested this embed in a single line.

Reply:
@igormcoelho

Reply:
And where is `p => knownHashes.Add(p)` part? hahahah

Reply:
IO caching Fifoset